packages/modules/vehicles/polestar/api.py

```diff
@@ -0,0 +1,106 @@
+            self.auth.delete_token()
+            log.error("query_params error: get response %d: unauthorized Exception", result.status_code)
+            return None
+        if result.status_code != 200:
+            log.error("query_params error: get response %d: %s", result.status_code, result.text)
+            return None
+
+        result_data = result.json()
+        if result_data.get('errors'):
+            error_message = result_data['errors'][0]['message']
+            log.error("query_params error: %s", error_message)
+            return None
+
+        log.debug(result_data)
+        return result_data
+
+    def get_battery_data(self) -> dict or None:
+        params = {
+            "query": "query GetBatteryData($vin: String!) { getBatteryData(vin: $vin) { "
+            + "batteryChargeLevelPercentage  estimatedDistanceToEmptyKm } }",
+            "operationName": "GetBatteryData",
+            "variables": "{\"vin\":\"" + self.vin + "\"}"
+        }
+
+        result = self.query_params(params)
+
+        if result is not None and result['data'] is not None and result['data']['getBatteryData'] is not None \
+                and result['data']['getBatteryData']['batteryChargeLevelPercentage'] is not None:
+            return result['data']['getBatteryData']
+        elif self.auth.access_token is not None:
+            # if we got an access code but the query failed, VIN could be wrong, so let`s check it
+            self.check_vin()
+            return None
+
+    def check_vin(self) -> None:
+        # get Vehicle Data
+        params = {
+            "query": "query GetConsumerCarsV2 { getConsumerCarsV2 { vin internalVehicleIdentifier __typename }}",
+            "operationName": "GetConsumerCarsV2",
+            "variables": "{}"
+        }
+        result = self.query_params(params, url='https://pc-api.polestar.com/eu-north-1/my-star/')
+        if result is not None and result['data'] is not None:
+            vins = []
+            # get list of cars and store the ones not matching our vin
+            cars = result['data']['getConsumerCarsV2']
+            if len(cars) == 0:
+                log.error("No vehicles found using specified account. Did you use the Polestar owners account?")
+                return None
+            for i in range(0, len(cars)):
+                if cars[i]['vin'] == self.vin:
+                    pass
+                else:
+                    vins.append(cars[i]['vin'])
+            if len(vins) > 0:
+                log.error("You probably specified a wrong VIN. We only found:%s", ",".join(vins))
+            return None
+
+
+def fetch_soc(user_id: str, password: str, vin: str, vehicle: int) -> CarState:
+    api = PolestarApi(user_id, password, vin)
+    bat_data = api.get_battery_data()
+    if bat_data is not None:
+        soc = bat_data['batteryChargeLevelPercentage']
+        est_range = bat_data['estimatedDistanceToEmptyKm']
+    else:
+        soc = 0
+        est_range = 0
```

Comment:
```suggestion
        soc = bat_data['batteryChargeLevelPercentage']
        est_range = bat_data['estimatedDistanceToEmptyKm']
```
Das zurücksetzen bei misslungener Abfrage erfolgt zentral.

Reply:
Verstanden